src/java/com/google/cloud/bigquery/dwhassessment/hooks/MigrationAssessmentLoggingHook.java

```diff
@@ -0,0 +1,130 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.QueryPlan;
+import org.apache.hadoop.hive.ql.exec.Utilities;
+import org.apache.hadoop.hive.ql.exec.mr.ExecDriver;
+import org.apache.hadoop.hive.ql.exec.tez.TezTask;
+import org.apache.hadoop.hive.ql.hooks.Entity;
+import org.apache.hadoop.hive.ql.hooks.ExecuteWithHookContext;
+import org.apache.hadoop.hive.ql.hooks.HookContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Hook, which captures query events and stores them as {@link GenericRecord}. */
+public class MigrationAssessmentLoggingHook implements ExecuteWithHookContext {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(MigrationAssessmentLoggingHook.class.getName());
+
+  private static final Schema QUERY_EVENT_SCHEMA = AvroSchemaLoader.loadSchema("QueryEvents.avsc");
+
+  // Just for the example. Real hook will write it to the file
+  public ArrayList<GenericRecord> records = new ArrayList<>();
+
+  public void run(HookContext hookContext) throws Exception {
+    processQueryEventAvro(hookContext);
+  }
+
+  private void processQueryEventAvro(HookContext hookContext) {
+    QueryPlan plan = hookContext.getQueryPlan();
+
+    LOG.info("Received hook notification for: " + plan.getQueryId());
+
+    // Make a copy so that we do not modify hookContext conf.
+    HiveConf conf = new HiveConf(hookContext.getConf());
+    List<ExecDriver> mrTasks = Utilities.getMRTasks(plan.getRootTasks());
+    List<TezTask> tezTasks = Utilities.getTezTasks(plan.getRootTasks());
+    ExecutionMode executionMode = getExecutionMode(mrTasks, tezTasks);
+
+    GenericRecord queryEvent =
+        new GenericRecordBuilder(QUERY_EVENT_SCHEMA)
+            .set("QueryId", plan.getQueryId())
+            .set("QueryText", plan.getQueryStr())
+            .set("EventType", hookContext.getHookType().name())
+            .set("Timestamp", plan.getQueryStartTime())
+            .set("User", getUser(hookContext))
+            .set("RequestUser", getRequestUser(hookContext))
+            .set("ExecutionMode", executionMode.name())
+            .set("Queue", getQueueName(executionMode, conf))
+            .set("TablesRead", getTablesFromEntitySet(plan.getInputs()))
+            .set("TablesWritten", getTablesFromEntitySet(plan.getOutputs()))
+            .build();
+    records.add(queryEvent);
+    LOG.info("Processed record: " + queryEvent.toString());
+  }
+
+  private List<String> getTablesFromEntitySet(Set<? extends Entity> entities) {
+    List<String> tableNames = new ArrayList<>();
+    for (Entity entity : entities) {
+      if (entity.getType() == TABLE || entity.getType() == PARTITION) {
+        tableNames.add(entity.getTable().getDbName() + "." + entity.getTable().getTableName());
```

Comment:
Optional: I haven't tested, but it's possible the expression:

```
entity.getTable().getDbName() + "." + entity.getTable().getTableName()
```

can be simplified to:

```
entity.getTable().getCompleteName()
```

Reply:
Done, but this part will be changed most likely anyway, since ProtoLoggingHook doesn't distinguish accessed partitions.